sklearn/tree/_splitter.pyx

```diff
@@ -236,6 +238,20 @@ cdef class Splitter:
 
         return self.criterion.node_impurity()
 
+    cdef inline bint check_monotonicity(self, INT32_t monotonic_cst, double lower_bound, double upper_bound) nogil:
+        """Check monotonic constraint is satisfied at the current split"""
+        cdef bint check_lower, check_upper, check_monotonic
+        check_lower = (self.criterion.sum_left[0] >= lower_bound * self.criterion.weighted_n_left) \
+                      & (self.criterion.sum_right[0] >= lower_bound * self.criterion.weighted_n_right)
+        check_upper = (self.criterion.sum_left[0] <= upper_bound * self.criterion.weighted_n_left) \
+                      & (self.criterion.sum_right[0] <= upper_bound * self.criterion.weighted_n_right)
+        if monotonic_cst == 0: # No constraint
+            return check_lower & check_upper
```

Comment:
why check anything when there's no constraint?

Reply:
Indeed, it's not needed.

Reply:
Actually I take it back, it's always needed to check the bounds !
Otherwise, with 2 features and a `POS` constraint on feature `0` the following tree can be accepted: 
```
                     X[0] <= 0
                     value = 100
                    /            \
         X[0] <= -1                X[1] <= 0
         value = 50                value = 150
       /            \             /            \
   leaf           leaf           leaf          leaf
   value = 25     value = 75     value = 50    value = 250
```
and the prediction would not be increasing with respect to `X[0]` as `predict([-0.5, -1]) = 75` and `predict([1, -1]) = 50`.

Reply:
Indeed.

Reply:
It would be interesting to come up with a synthetic dataset that would lead to learning this tree without the constraint on `X[0]` and check in a test that this does not happen when the constraint is enabled.

The same test should be reused for `HistGradientBoostingRegressor`.

Reply:
Or maybe this is already covered by `test_nd_tree_nodes_values`? I am not sure.

Reply:
This is not covered by `test_nd_tree_nodes_values`, which only tests that such a tree doesn't occur while enforcing the constraint, and not that it does occurs while not enforcing the constraint.

Reply:
4e3edac implements clipping and removes this check.